### PR TITLE
feat/point in time recovery

### PR DESCRIPTION
**Pull request by joshenlim**
### UI for point in time recovery (backups)

- Feature is toggleable via feature flag, for controlled release, in which if hidden it'll show a "Coming soon" state

Remaining to-dos once BE is completed
- [x] Hook up endpoint to call physical restore
- [x] E2E testing: ensure that existing UI handles restoring state to being back online and verify correctness

Nice to have:
- [ ] Use a better DatePicker component (may need to update the component to support time input, non-range input etc)

https://user-images.githubusercontent.com/19742402/172806687-af1468fd-fec1-4522-a61e-32bd14dd4750.mov

### CODE REVIEW

**Comment by saltcod** [comment]
@joshenlim at last!   This looks great!  Couple of small UI things I thought of while playing with this.

(1)
CleanShot 2022-06-23 at 15 19 10@2x


I played around in Sketch a little and started down this road a little:
CleanShot 2022-06-23 at 15 21 00@2x


But wonder if that's just unnecessary additional ui. 
Thinking separating the 3 fields + adding some simple helper text telling you to pick a date could be sufficient: 

CleanShot 2022-06-23 at 15 25 45@2x

And / or changing the earliest/latest input fields to look less like input fields: 

CleanShot 2022-06-23 at 15 25 31@2x


(2)
Wonder if it would be useful to clarify a little how point-in-time recoveries work. How often are they made? If one is made at 1:05pm and I select 1:03pm, what happens then?  A quick explanation might help there:

CleanShot 2022-06-23 at 15 40 51@2x


(3)
Last, the input error state icon is overlapping the calendar icon here:
CleanShot 2022-06-23 at 15 08 27@2x

(4)
On the language front there — I wonder if "Oldest backup / Newest Backup" is clearer than "Earliest point of recovery" and if it conveys the same info.  

:raised_hands:

File: studio/components/interfaces/Database/Backups/BackupsEmpty.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { IconInfo } from '@supabase/ui'
+
+const BackupsEmpty = () => {
+  return (
+    <div className="block w-full rounded border border-gray-400 border-opacity-50 bg-gray-300 p-3">
+      <div className="flex space-x-3">
+        <IconInfo size={20} strokeWidth={1.5} />
+        <p className="text-sm">No backups created yet. Check again tomorrow.</p>
```
Comment (saltcod) [inline studio/components/interfaces/Database/Backups/BackupsEmpty.tsx:8]: I found this a little confusing when I was browsing my test project. Do we know when backups are done? Nightly SG time or some other time? Would be nice to provide any extra info we have there.

**Comment by joshenlim** [comment]
Gotcha on these - i'll probably just go with separating the 3 fields, that makes sense too since only the recovery point input is what's important

RE: clarify a little how point-in-time recoveries work.
this one's a little too complex to try a simplified explanation tbh - our backend implementation is based on [this](https://github.com/wal-g/wal-g), but TLDR users can literally choose _any_ point in time within the range, we restore it back to that exact time, down to the seconds

RE: error input state
ahh this one's a UI bug on the ui library i think (cc @MildTomato )

File: studio/components/interfaces/Database/Backups/BackupsEmpty.tsx (hunk 1 above)
Comment (joshenlim) [reply studio/components/interfaces/Database/Backups/BackupsEmpty.tsx:8]: checking with angelico on the details for scheduled backups

**Comment by joshenlim** [comment]
RE: earliest and latest not using input components
main consideration here was so that the recovery point input field would line up with the earliest and latest points of recovery text - easier on the eyes as a comparison to check your input to make sure its within the range

**Comment by joshenlim** [comment]
Updated the UI a little:

File: studio/components/interfaces/Database/Backups/BackupsEmpty.tsx (hunk 1 above)
Comment (joshenlim) [reply studio/components/interfaces/Database/Backups/BackupsEmpty.tsx:8]: scheduled backups happen at 12 midnight based on the project's region - i'll add that message here to make it clearer

Comment (joshenlim) [reply studio/components/interfaces/Database/Backups/BackupsEmpty.tsx:8]: image

Comment (saltcod) [reply studio/components/interfaces/Database/Backups/BackupsEmpty.tsx]: Nice!

**Comment by saltcod** [comment]
Nice! I think that's enough. Feels much better with that small change.

**Comment by joshenlim** [comment]
@saltcod could i get a +1? i'll get this in on monday

File: studio/components/interfaces/Database/Backups/BackupsList.tsx [hunk 2]
```diff
@@ -1,48 +1,28 @@
...
+import BackupItem from './BackupItem'
+import BackupsError from './BackupsError'
+import BackupsEmpty from './BackupsEmpty'
+
 interface Props {}
 
 const BackupsList: FC<Props> = ({}) => {
-  const { ui, app } = useStore()
+  const { ui, backups } = useStore()
   const projectRef = ui.selectedProject?.ref || 'default'
```
Comment (alaister) [inline studio/components/interfaces/Database/Backups/BackupsList.tsx:17]: Instance 1

File: studio/components/interfaces/Database/Backups/PITRBackupSelection/index.tsx [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  const [searchString, setSearchString] = useState<string>('')
+  const [selectedTimezone, setSelectedTimezone] = useState<any>(getClientTimezone())
+  const [showConfirmation, setShowConfirmation] = useState<boolean>(false)
+
+  const [errors, setErrors] = useState<any>()
+  const [recoveryPoint, setRecoveryPoint] = useState<string>()
+
+  const { configuration } = backups
+  const projectId = ui.selectedProject?.id ?? -1
+  const projectRef = ui.selectedProject?.ref ?? 'default'
```
Comment (alaister) [inline studio/components/interfaces/Database/Backups/PITRBackupSelection/index.tsx:46]: Instance 2

File: studio/pages/project/[ref]/database/backups/pitr.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { useStore } from 'hooks'
+import { DatabaseLayout } from 'components/layouts'
+import { PITRBackupSelection } from 'components/interfaces/Database'
+import { NextPageWithLayout } from 'types'
+
+const DatabaseScheduledBackups: NextPageWithLayout = () => {
+  const { ui } = useStore()
+  const router = useRouter()
+
+  const ref = ui.selectedProject?.ref ?? 'default'
```
Comment (alaister) [inline studio/pages/project/[ref]/database/backups/pitr.tsx:14]: Instance 3

**Review by alaister — COMMENTED** [review]
What do you think of changing `'default'` to `'_'` so if the user does somehow manage to click a link before the `projectRef` is set, they'll just be taken to the project selection screen?
I think I've highlighted all the instances where `'default'` is used below.

**Comment by joshenlim** [comment]
@alaister 
> changing 'default' to '_'

i've made it `default` to support local studio actually - the project ref for local studio is `default`

**Review by alaister — APPROVED** [review]
Nice work @joshenlim!

**Pull request merged**